target/cheri-common/cheri_utils.h

```diff
@@ -133,10 +133,19 @@ static inline cap_length_t cap_get_top_full(const cap_register_t *c)
 
 static inline bool cap_otype_is_reserved(target_ulong otype)
 {
+    cheri_debug_assert((target_long)otype >= 0 &&
+                       "Should not only be used with sign-extended otypes!");
+    cheri_debug_assert(otype <= CAP_MAX_REPRESENTABLE_OTYPE &&
+                       "Should only be called for in-range otypes!");
     return otype >= CAP_CC(FIRST_SPECIAL_OTYPE) &&
            otype <= CAP_CC(LAST_SPECIAL_OTYPE);
 }
 
+static inline bool cap_otype_is_nonreserved(target_ulong otype)
+{
+    return otype < CAP_MAX_REPRESENTABLE_OTYPE && !cap_otype_is_reserved(otype);
```

Comment:
<=? Otherwise max otype isn't handled correctly on Morello (but happens to be on CHERI due to that being a reserved otype)

Reply:
Also, cap_otype_is_reserved asserting but cap_otype_is_reserved not asserting is a bit confusing, it's really tempting to turn the !cap_otype_is_nonreserved calls into cap_otype_is_reserved, but can't due to the difference between the two